packages/zoe/README.md

```diff
@@ -51,3 +51,60 @@ agoric follow :published.priceFeed.ATOM-USD_price_feed
 ```
 
  TODO document more of https://github.com/Agoric/documentation/issues/672
+
+## Upgrade
+
+A contract instance can be upgraded to use a new source code bundle in a process that is very similar to
+[upgrading a vat](https://github.com/Agoric/agoric-sdk/blob/master/packages/SwingSet/docs/vat-upgrade.md).
+Most state is discarded, however "durable" collections are retained for use by the replacement version.
+
+The upgrade process is triggered through the "adminFacet" of the instance, and requires specifying the new source bundle. (Note that a "null upgrade" that re-uses the original bundle is valid, and a legitimate approach to deleting accumulated state).
+
+```js
+const results = E(instanceAdminFacet).upgradeContract(newBundleId);
+```
+
+The new bundle itself must export a `prepare` function in place of `start`, and is obligated to redefine every durable Kind that was created by its predecessor.
+
+```js
+const prepare = async (zcf, _privateArgs, instanceBaggage) => {
+  const isFirstVersion = !instanceBaggage.has('version');
+  if (isFirstVersion) {
+    instanceBaggage.init('version', 1);
+  } else {
+    const previousVersion = instanceBaggage.get('version');
+    instanceBaggage.set('version', previousVersion + 1);
+  }
+
+  const CounterI = M.interface('Counter', {
+    increment: M.call().returns(M.bigint()),
+    read: M.call().returns(M.bigint()),
+  });
+  const initCounterState = () => ({ value: 0n });
+  const makeCounter = prepareExoClass(
+    instanceBaggage,
+    'Counter',
+    CounterI,
+    initCounterState,
+    {
+      increment() { return this.state.value += 1n; },
```

Comment:
As discussed in the show-and-tell, we should have an upgrade example somewhere that shows the v1 code and v2 code separately, with another section that specifically shows a diff. Teaching upgrade is always kinda tricky, but I think an example where you add a new method would help make things more concrete. Especially if it tells a story like "you deployed v1 in March, and then in April you realized you wanted a new feature, here's how you roll it out". It's too easy for a single compact example to tell the end state without describing the starting state.

Reply:
I agree. The example should also show how the v2 contract triggers different behaviors based on whether the `'version'` in `instanceBaggage` indicates an upgrade is needed or not.

Reply:
Done.